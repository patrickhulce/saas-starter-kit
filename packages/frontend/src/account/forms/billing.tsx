import * as React from 'react'
import {IUser} from '../../../../shared/lib/typedefs'

export class BillingForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element[] {
    return [
      <h2>Billing</h2>,
    ]
  }
}
