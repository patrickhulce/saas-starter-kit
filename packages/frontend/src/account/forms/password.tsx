import * as React from 'react'

import {IUser} from '../../../../shared/lib/typedefs'

export class PasswordForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element[] {
    return [
      <h2 key="password">Password</h2>,
    ]
  }
}
