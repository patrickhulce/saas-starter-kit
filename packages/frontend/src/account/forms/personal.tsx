import Button from '@material-ui/core/Button'
import TextField from '@material-ui/core/TextField'
import * as React from 'react'

import {IUser} from '../../../../shared/lib/typedefs'
import {ErrorBar} from '../../components/error-bar/error-bar'
import {Form, IFormData} from '../../components/form'
import {LoadingBar} from '../../components/loading-bar/loading-bar'

async function updateAccount(user: IUser): Promise<void> {
  const response = await fetch(`/api/v1/users/${user.id}`, {
    method: 'PUT',
    body: JSON.stringify(user),
    headers: {'content-type': 'application/json'},
  })

  if (response.status !== 200) {
    // TODO: Provide more specific error message
    throw new Error('Error updating account')
  }
}

interface INamesFormState {
  isLoading?: boolean
  errorMessage?: string
}

class NamesForm extends Form<{user: IUser}, INamesFormState> {
  public state: INamesFormState = {}

  protected async _handleSubmit(data: IFormData): Promise<void> {
    const user: IUser = {...this.props.user, firstName: data.first, lastName: data.last}
    await updateAccount(user)
  }

  public render(): JSX.Element {
    return (
      <form onSubmit={this.handleSubmit}>
        <LoadingBar isLoading={this.state.isLoading} />
        <ErrorBar message={this.state.errorMessage} />
        <TextField
          id="first-name"
          name="first"
          label="First Name"
          defaultValue={this.props.user.firstName}
          required
        />
        <TextField
          id="last-name"
          name="last"
          label="Last Name"
          defaultValue={this.props.user.lastName}
          required
        />
        <Button variant="raised" color="primary" type="submit">
          Update
        </Button>
      </form>
    )
  }
}

export class PersonalForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element[] {
    return [
      <h2 key="header">Personal Information</h2>,
      <NamesForm key="names-form" user={this.props.user} />,
    ]
  }
}
