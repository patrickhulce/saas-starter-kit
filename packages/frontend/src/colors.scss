$primary-brand: #0771b8;
$accent-brand: #fcb827;

$darker-gray: #333;
$dark-gray: #444;
$medium-gray: #666;
$lightest-gray: #eee;

$dark-red: #721c24;
$light-pink: #f5c6c6;
$lightest-pink: #fff4f4;

$dark-green: #1c721c;
$light-green: #c6f5cd;
$lightest-green: #f5fff4;
