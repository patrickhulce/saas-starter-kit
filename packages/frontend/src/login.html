<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>THE_PRODUCT_DISPLAY_NAME Login</title>
</head>

<body>
  <h1 class="app-name">THE_PRODUCT_DISPLAY_NAME</h1>
  <div id="app-root"></div>
</body>

</html>
