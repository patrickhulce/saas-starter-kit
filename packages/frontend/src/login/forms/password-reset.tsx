import * as React from 'react'

import {BasicTextField} from '../../components/basic-text-field'
import {Form, IFormData, IFormState, ISubmitOptions} from '../../components/form/form'
import {resetPassword} from '../../services/user-service'

export class PasswordResetForm extends Form {
  public state: IFormState = {}
  public submitUIOptions: ISubmitOptions = {label: 'Reset Password'}

  protected async _handleSubmit(data: IFormData): Promise<void> {
    if (data.confirmPassword !== data.password) {
      throw new Error('Passwords did not match')
    }

    const key = new URLSearchParams(window.location.search).get('reset-password-key')
    if (!key) throw new Error('Password reset key expired')
    await resetPassword(key, data.password)
  }

  public renderInputUI(): JSX.Element {
    return (
      <React.Fragment>
        <BasicTextField name="password" type="password" autoFocus />
        <BasicTextField name="confirmPassword" type="password" autoFocus />
      </React.Fragment>
    )
  }
}
