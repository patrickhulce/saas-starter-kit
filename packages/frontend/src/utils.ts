import * as React from 'react'
import * as ReactDOM from 'react-dom'

import {IUser} from '../../shared/lib/typedefs'

import {getLoggedInUser, refreshLoggedInUser} from './services/user-service'

export async function redirectToLogin(): Promise<void> {
  window.location.href = '/login'
  return new Promise(resolve => setTimeout(resolve, 10000)) as Promise<void>
}

export async function redirectIfNotAuthorized(): Promise<void> {
  try {
    const localUser = await getLoggedInUser()
    if (!localUser) await redirectToLogin()
    await refreshLoggedInUser()
    const remoteUser = await getLoggedInUser()
    if (!remoteUser || localUser!.id !== remoteUser.id) await redirectToLogin()
  } catch (err) {
    await redirectToLogin()
  }
}

export async function findUserOnStartupOrBail(): Promise<IUser> {
  const localUser = await getLoggedInUser()
  if (!localUser) await redirectToLogin()
  setTimeout(redirectIfNotAuthorized, 0)
  return localUser!
}

export function createRenderFn(render: () => React.ReactElement<any>): () => void {
  return () => ReactDOM.render(render(), document.getElementById('app-root'))
}

export function HMR(nodeModule: any, handle: (nodeModule: any) => any): void {
  if (nodeModule.hot) {
    handle(nodeModule)
  }
}

export const testIds = {
  loadingBar: 'loading-bar',
  messageBar: 'message-bar',
  // Login IDs
  createAccountTab: 'create-account-tab',
  createAccountSubmit: 'create-account-submit',
  loginFormPrimaryBtn: 'login-form-primary-btn',
  loginFormSecondaryBtn: 'login-form-secondary-btn',
  loginForm: 'login-form',
  registerForm: 'register-form',
  profileNamesForm: 'profile-names-form',
}
