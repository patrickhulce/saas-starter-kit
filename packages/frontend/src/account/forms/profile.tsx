import * as React from 'react'

import {IUser} from '../../../../shared/lib/typedefs'
import {BasicTextField} from '../../components/basic-text-field'
import {Form, IFormData, IFormState} from '../../components/form/form'
import * as formStyles from '../../components/form/form.scss'
import {updateAccount} from '../../services/user-service'
import {testIds} from '../../utils'

class NamesForm extends Form<{user: IUser}> {
  public state: IFormState = {}
  public testId: string = testIds.profileNamesForm

  protected async _handleSubmit(data: IFormData): Promise<void> {
    await updateAccount(this.props.user.id, {firstName: data.firstName, lastName: data.lastName})
  }

  public renderInputUI(): JSX.Element {
    return (
      <React.Fragment>
        <BasicTextField name="firstName" defaultValue={this.props.user.firstName} />
        <BasicTextField name="lastName" defaultValue={this.props.user.lastName} />
      </React.Fragment>
    )
  }
}

export class ProfileForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element {
    const formClasses = `${formStyles.columnarForm} ${formStyles.fixedWidth}`
    return (
      <React.Fragment>
        <h2>Personal Information</h2>
        <NamesForm user={this.props.user} className={formClasses} />
      </React.Fragment>
    )
  }
}
