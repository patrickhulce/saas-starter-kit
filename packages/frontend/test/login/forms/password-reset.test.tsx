import * as React from 'react'
import {RenderResult, fireEvent, render, wait} from 'react-testing-library'

import {PasswordResetForm} from '../../../src/login/forms/password-reset'
import {testIds} from '../../../src/utils'
import {createFetchPromise} from '../../utils'

describe('register/forms/register.tsx', () => {
  let fetchMock: jest.Mock
  let locationAssignMock: jest.Mock

  function renderAndFill(): RenderResult {
    const form = render(<PasswordResetForm />)
    const password = form.getByLabelText(/^Password/) as HTMLInputElement
    const confirmPassword = form.getByLabelText(/Confirm Password/) as HTMLInputElement
    password.value = 'new_password'
    confirmPassword.value = 'new_password'
    return form
  }

  beforeEach(() => {
    fetchMock = jest.fn()
    self.fetch = fetchMock
    // mock the URL, see https://www.ryandoll.com/post/2018/3/29/jest-and-url-mocking
    window.history.pushState({}, '', '/login?reset-password-key=1234key1234')
    window.location.assign = locationAssignMock = jest.fn()
  })

  it('should render form', () => {
    const form = render(<PasswordResetForm />)
    expect(form.container).toMatchSnapshot()
  })

  it('should check passwords match', async () => {
    const {getByTestId, getByLabelText} = renderAndFill()
    const confirmPassword = getByLabelText(/Confirm Password/) as HTMLInputElement
    confirmPassword.value = 'other'

    fireEvent.submit(getByTestId(testIds.passwordResetForm))
    await wait(() => getByTestId(testIds.messageBar))

    expect(getByTestId(testIds.messageBar).textContent).toMatchSnapshot()
  })

  it('should send password reset request to server', async () => {
    const {getByTestId} = renderAndFill()
    const fetchPromise = createFetchPromise({status: 204})
    fetchMock.mockImplementation(fetchPromise.fn)

    fireEvent.submit(getByTestId(testIds.passwordResetForm))
    await wait(() => getByTestId(testIds.loadingBar))

    fetchPromise.resolve()
    expect(fetchMock).toHaveBeenCalled()
    expect(fetchMock.mock.calls[0]).toMatchSnapshot()
    await wait(() => expect(locationAssignMock).toHaveBeenCalledWith('/login'))
  })

  it('should show an error UI', async () => {
    const {getByTestId, queryByTestId} = renderAndFill()
    const fetchPromise = createFetchPromise({status: 400})
    fetchMock.mockImplementation(fetchPromise.fn)

    fireEvent.submit(getByTestId(testIds.passwordResetForm))
    await wait(() => getByTestId(testIds.loadingBar))

    expect(queryByTestId(testIds.messageBar)).toBeNull()

    fetchPromise.resolve()
    await wait(() => getByTestId(testIds.messageBar))

    expect(queryByTestId(testIds.loadingBar)).toBeNull()
    expect(getByTestId(testIds.messageBar).textContent).toMatchSnapshot()
  })
})
