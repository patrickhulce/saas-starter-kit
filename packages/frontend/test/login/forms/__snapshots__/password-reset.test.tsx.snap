// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register/forms/register.tsx should check passwords match 1`] = `"Passwords did not match"`;

exports[`register/forms/register.tsx should render form 1`] = `
<div>
  <form
    data-testid="password-reset-form"
  >
    <div
      class="MuiFormControl-root-1"
    >
      <label
        class="MuiFormLabel-root-10 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
        data-shrink="false"
        for="password-input"
      >
        Password
        <span
          class="MuiFormLabel-asterisk-14"
        >
           *
        </span>
      </label>
      <div
        class="MuiInput-root-15 MuiInput-formControl-16 MuiInput-underline-19"
      >
        <input
          aria-invalid="false"
          aria-required="true"
          class="MuiInput-input-23 MuiInput-inputType-26"
          id="password-input"
          name="password"
          required=""
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-1"
    >
      <label
        class="MuiFormLabel-root-10 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
        data-shrink="false"
        for="confirm-password-input"
      >
        Confirm Password
        <span
          class="MuiFormLabel-asterisk-14"
        >
           *
        </span>
      </label>
      <div
        class="MuiInput-root-15 MuiInput-formControl-16 MuiInput-underline-19"
      >
        <input
          aria-invalid="false"
          aria-required="true"
          class="MuiInput-input-23 MuiInput-inputType-26"
          id="confirm-password-input"
          name="confirmPassword"
          required=""
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      class="MuiButtonBase-root-50 MuiButton-root-28 MuiButton-contained-37 MuiButton-containedPrimary-38 MuiButton-raised-40 MuiButton-raisedPrimary-41"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label-29"
      >
        Reset Password
      </span>
      <span
        class="MuiTouchRipple-root-53"
      />
    </button>
  </form>
</div>
`;

exports[`register/forms/register.tsx should send password reset request to server 1`] = `
Array [
  "/api/v1/users/password-reset",
  Object {
    "body": "{\\"key\\":\\"1234key1234\\",\\"password\\":\\"new_password\\"}",
    "headers": Object {
      "content-type": "application/json",
    },
    "method": "PUT",
  },
]
`;

exports[`register/forms/register.tsx should show an error UI 1`] = `"Password reset failed"`;
